x509/src/pkix_extensions.rs

```diff
@@ -0,0 +1,1041 @@
+/// - [`PKIX_CE_ANYEXTENDEDKEYUSAGE`](constant.PKIX_CE_ANYEXTENDEDKEYUSAGE.html),
+/// - [`PKIX_KP_SERVERAUTH`](constant.PKIX_KP_SERVERAUTH.html),
+/// - [`PKIX_KP_CLIENTAUTH`](constant.PKIX_KP_CLIENTAUTH.html),
+/// - [`PKIX_KP_CODESIGNING`](constant.PKIX_KP_CODESIGNING.html),
+/// - [`PKIX_KP_EMAILPROTECTION`](constant.PKIX_KP_EMAILPROTECTION.html),
+/// - [`PKIX_KP_TIMESTAMPING`](constant.PKIX_KP_TIMESTAMPING.html),
+///
+/// ```text
+/// ExtKeyUsageSyntax ::= SEQUENCE SIZE (1..MAX) OF KeyPurposeId
+/// ```
+///
+/// [RFC 5280 Section 4.2.1.12]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.12
+pub type ExtendedKeyUsage<'a> = alloc::vec::Vec<ObjectIdentifier>;
+
+/// Subject alternative name extension as defined in [RFC 5280 Section 4.2.1.6] and as identified by the [`PKIX_CE_SUBJECT_ALT_NAME`](constant.PKIX_CE_SUBJECT_ALT_NAME.html) OID.
+///
+/// ```text
+/// SubjectAltName ::= GeneralNames
+/// ```
+///
+/// [RFC 5280 Section 4.2.1.6]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6
+pub type SubjectAltName<'a> = GeneralNames<'a>;
+
+/// Issuer alternative name extension as defined in [RFC 5280 Section 4.2.1.7] and as identified by the [`PKIX_CE_ISSUER_ALT_NAME`](constant.PKIX_CE_ISSUER_ALT_NAME.html) OID.
+///
+/// ```text
+/// IssuerAltName ::= GeneralNames
+/// ```
+///
+/// [RFC 5280 Section 4.2.1.7]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.7
+pub type IssuerAltName<'a> = GeneralNames<'a>;
+
+/// OCSP noCheck extension as defined in [RFC 6960 Section 4.2.2.2.1] and as idenfied by the [`PKIX_OCSP_NOCHECK`](constant.PKIX_OCSP_NOCHECK.html) OID.
+///
+/// ```text
+/// OcspNoCheck ::= NULL
+/// ```
+///
+/// [RFC 6960 Section 4.2.2.2.1]: https://datatracker.ietf.org/doc/html/rfc6960#section-4.2.2.2.1
+pub type OcspNoCheck = Null;
+
+/// Subject directory attributes extension as defined in [RFC 5280 Section 4.2.1.8] and as identified by the [`PKIX_CE_SUBJECT_DIRECTORY_ATTRIBUTES`](constant.PKIX_CE_SUBJECT_DIRECTORY_ATTRIBUTES.html) OID.
+///
+/// ```text
+/// SubjectDirectoryAttributes ::= SEQUENCE SIZE (1..MAX) OF AttributeSet
+/// ```
+///
+/// [RFC 5280 Section 4.2.1.8]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.8
+pub type SubjectDirectoryAttributes<'a> = alloc::vec::Vec<AttributeTypeAndValue<'a>>;
+
+/// Basic constraints extension as defined in [RFC 5280 Section 4.2.1.9] and as identified by the [`PKIX_CE_BASIC_CONSTRAINTS`](constant.PKIX_CE_BASIC_CONSTRAINTS.html) OID.
+///
+/// ```text
+///    BasicConstraints ::= SEQUENCE {
+///         cA                      BOOLEAN DEFAULT FALSE,
+///         pathLenConstraint       INTEGER (0..MAX) OPTIONAL }
+/// ```
+///
+/// [RFC 5280 Section 4.2.1.9]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+pub struct BasicConstraints {
+    /// cA                      BOOLEAN DEFAULT FALSE,
+    #[asn1(default = "default_false")]
```

Comment:
Curious if this "just works" as an alternative:

```suggestion
    #[asn1(default = "Default::default")]
```

If so, you could use that pattern as an alternative to having to write a custom function that returns the `Default` for a given type, with no additional changes to the proc macro.

Reply:
That did just work for here. Still have not added anything with default alg ID yet (CMS support will generate those).